icu4j/main/core/src/main/java/com/ibm/icu/message2/MFParser.java

```diff
@@ -46,22 +46,38 @@ public static MFDataModel.Message parse(String input) throws MFParseException {
     // Parser proper
     private MFDataModel.Message parseImpl() throws MFParseException {
         MFDataModel.Message result;
+        // Determine if message is simple or complex; this requires
+        // looking through whitespace.
+        int savedPosition = input.getPosition();
+        while (true) {
+            int cp = input.peekChar();
+            if (!StringUtils.isWhitespace(cp)) {
+                break;
+            }
+            // Skip past whitespace
+            cp = input.readCodePoint();
+        }
         int cp = input.peekChar();
         if (cp == '.') { // declarations or .match
+            // No need to restore whitespace
             result = getComplexMessage();
         } else if (cp == '{') { // `{` or `{{`
             cp = input.readCodePoint();
             cp = input.peekChar();
             if (cp == '{') { // `{{`, complex body without declarations
                 input.backup(1); // let complexBody deal with the wrapping {{ and }}
                 MFDataModel.Pattern pattern = getQuotedPattern();
+                skipOptionalWhitespaces();
```

Comment:
I think it would really help to have the ebnf in a comment above this.
To show if there should be some optional whitespaces there or not.

Reply:
Done in [c14417b](https://github.com/unicode-org/icu/pull/3063/commits/c14417b9c098b383f1a5134a99456863cfdb3f91)